src/librustc_middle/ty/layout.rs

```diff
@@ -1141,6 +1033,194 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
+                                );
+                                niche.offset
+                            } else {
+                                st[dataful_variant].fields.offset(field_index) + niche.offset
+                            };
+
+                            let new_size = st[dataful_variant].size.align_to(align.abi);
+                            debug_assert!(
+                                !struct_reordering_opt || new_size == max_size.align_to(align.abi)
+                            );
+                            debug_assert!(st.iter().all(|v| { v.size <= new_size }));
+
+                            let new_largest_niche =
+                                Niche::from_scalar(dl, niche_offset, niche_scalar.clone());
+
+                            if new_size < size
+                                || new_largest_niche.as_ref().map_or(0, |n| n.available(dl))
+                                    > largest_niche.as_ref().map_or(0, |n| n.available(dl))
+                            {
+                                let abi = if st.iter().all(|v| v.abi.is_uninhabited()) {
+                                    Abi::Uninhabited
+                                } else if second_max_size == Size::ZERO {
+                                    match st[dataful_variant].abi {
+                                        Abi::Scalar(_) => Abi::Scalar(niche_scalar.clone()),
+                                        Abi::ScalarPair(ref first, ref second) => {
+                                            // We need to use scalar_unit to reset the
+                                            // valid range to the maximal one for that
+                                            // primitive, because only the niche is
+                                            // guaranteed to be initialised, not the
+                                            // other primitive.
+                                            if niche_offset.bytes() == 0 {
+                                                Abi::ScalarPair(
+                                                    niche_scalar.clone(),
+                                                    scalar_unit(second.value),
+                                                )
+                                            } else {
+                                                Abi::ScalarPair(
+                                                    scalar_unit(first.value),
+                                                    niche_scalar.clone(),
+                                                )
+                                            }
+                                        }
+                                        _ => Abi::Aggregate { sized: true },
+                                    }
+                                } else {
+                                    Abi::Aggregate { sized: true }
+                                };
+
+                                return Ok(tcx.intern_layout(Layout {
```

Comment:
Btw you can do this without moving the code around by setting a `None` variable to `Some(tcx.intern_layout(Layout {...}))`, and only returning that after creating the `Layout` for the tagged layout.

But the current state of the PR should be enough for perf testing.

Reply:
Ah oops, I guess you didn't take my suggestion, hmm.